File: 108 Convert Sorted Array to Binary Search Tree.c
/*
Convert Sorted Array to Binary Search Tree 
https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/
*/
struct TreeNode
{
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};
typedef struct TreeNode node;

void getBST(node* n, int* arr, int lt, int rt)
{
    int m = ((lt + rt) >> 1);
    n->val = arr[m];
    
    if(lt <= m-1)
    {
        n->left  = (node*)malloc(sizeof(node));
        getBST(n->left, arr,lt,m-1);
    }
    else 
        n->left = 0;
    
    if(m+1 <= rt)
    {
        n->right = (node*)malloc(sizeof(node));
        getBST(n->right,arr,m+1,rt);
    }
    else
        n->right = 0;
}

node* sortedArrayToBST(int* arr, int size)
{
    node* n = 0;
    if(size < 1) return n;
    
    n = (node*)malloc(sizeof(node));
    getBST(n,arr,0,size-1);
    
    return n;
}

File: 108 Convert Sorted Array to Binary Search Tree.java
/*
108. Convert Sorted Array to Binary Search Tree
https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/

Given an array where elements are sorted in ascending order, convert it to a height balanced BST.

For this problem, a height-balanced binary tree is defined as a binary tree in which
the depth of the two subtrees of every node never differ by more than 1.

Example:

Given the sorted array: [-10,-3,0,5,9],

One possible answer is: [0,-3,9,-10,null,5], which represents the following height balanced BST:

      0
     / \
   -3   9
   /   /
 -10  5
*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode sortedArrayToBST(int[] nums) {
        if (nums.length < 1) {
            return null;
        }
        
        return inorder(nums, 0, nums.length - 1);
    }
    
    /**
    * left and right: indexes of nums, inclusive
    */
    private TreeNode inorder(int[] nums, int left, int right) {
        if (left > right) {
            return null;
        }
        
        int mid = left + ((right - left) >> 1);
        TreeNode node = new TreeNode(nums[mid]);
        
        node.left = inorder(nums, left, mid - 1);
        node.right = inorder(nums, mid + 1, right);
        
        return node;
    }
}
